### PR TITLE
Add argument parsing

### PR DESCRIPTION
**Pull request by Arignir**

### CODE REVIEW

**Comment by Arignir** [comment]
I just realised some parts of the PR wouldn't execute properly due to missing fields in `Build` and `Package`.
Honestly, it's quite hard to provide something that would run smoothly no matter what and keep the PR small, so i guess we'll ignore those issues until I push the content of `Build`, `Package` and `Manifest`, eh? 

(yeah, it sucks, but we leave in a horrible world. Wake up sheeple.)

File: core/args.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    )
+    nbuild_parser.add_argument(
+        '-o',
+        '--output-dir',
+        default=os.path.join(  # Default path is script_dir/packages
+            os.getcwd(),
+            os.path.dirname(sys.argv[0]),
+            'packages',
+        ),
+        help='Output directory for built packages. Default: packages/',
```
Comment (Pedrochu) [inline core/args.py]: Shouldn't those help messages be between double quotes instead of single ones?

File: core/args.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        '--purge',
+        action='store_true',
+        help='Remove all cached data.',
+    )
+    nbuild_parser.add_argument(
+        '-r',
+        '--repository',
+        default='stable',
+        metavar='REPOSITORY',
+        help='Name of the repository built packages will be a part of. Default: stable',
```
Comment (Pedrochu) [inline core/args.py]: I'm guessing it's missing a `the` before `built`

File: core/cache.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        get_args().cache_dir,
+        'build',
+        build.manifest.name,
+        build.semver,
+    )
+
+
+def get_wrap_cache(package: Package):
+    """
+    Returns the location of the cache where the files composing of the given package should be placed.
```
Comment (Pedrochu) [inline core/cache.py]: the `of` should be removed on `composing of the given package`

File: core/log.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    try:
+        yield
+    finally:
+        log_tab_level -= 1
+
+
+def dlog(*args, indent=True):
+    """
+    Prints a debug log, prefix by a magenta `[d]`.
```
Comment (Pedrochu) [inline core/log.py]: *prefixed

File: core/log.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    """
+    global log_tab_level
+
+    indent = '\t' * log_tab_level
+    print(f"{termcolor.colored('[!]', 'yellow', attrs=['bold'])} {indent}", *args)
+
+
+def elog(*args, indent=True):
+    """
+    Prints an non-fatal error log, prefixed by a red `[-]`.
```
Comment (Pedrochu) [inline core/log.py]: it's `a`, not `an`
there's the same error on the next function

File: core/log.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    indent = '\t' * log_tab_level
+    print(f"{termcolor.colored('[-]', 'red', attrs=['bold'])} {indent}", *args)
+
+
+def flog(args):
+    """
+    Prints an fatal error log, all in red, prefixed by `[-]`.
+    This function does NOT abort the current process's execution.
+    """
+    termcolor.cprint(f"[-] {args}", "red", attrs=['bold'])
```
Comment (Pedrochu) [inline core/log.py]: The `red` should be between `'`

File: nbuild.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    # Target & host architecture
+    # TODO FIXME Set as parameter
+    os.environ['TARGET'] = 'x86_64-linux-gnu'
+    os.environ['HOST'] = 'x86_64-linux-gnu'
+
+    # Common flags for the gnu toolchain (cpp, cc, cxx, as, ld)
+    gnuflags = '-O2 -s -m64 -mtune=generic'
+
+    # Pre-processor
+    os.environ['CPP'] = f'{os.environ["TARGET"]}-cpp'
```
Comment (Pedrochu) [inline nbuild.py:43]: Shouldn't `TARGET` should be between `'` instead of `"`? There's this error throughout this file, same goes for `HOST`

**Review by Pedrochu — CHANGES_REQUESTED** [review]
Found some typos and English mistakes, it shouldn't be too hard to fix

File: core/args.py (hunk 1 above)
Comment (Arignir) [reply core/args.py]: You're right :+1:

File: nbuild.py (hunk 7 above)
Comment (Arignir) [reply nbuild.py:43]: It is not possible to put it in between `'` because of how fstrings works. Yeah, it sucks :/

File: core/args.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            os.getcwd(),
+            os.path.dirname(sys.argv[0]),
+            'cache',
+        ),
+        help="Cache directory used when downloading and building packages. Default: cache/",
+    )
+    nbuild_parser.add_argument(
+        '--purge',
+        action='store_true',
+        help='Remove all cached data.',
```
Comment (Pedrochu) [inline core/args.py]: double quotes :stuck_out_tongue_winking_eye:

Comment (Arignir) [reply core/args.py]: Damn it :(

**Review by Pedrochu — DISMISSED** [review]

File: core/args.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        default=0,
+        help="Make the operation more talkative. Append it multiple times to make it even more talkative."
+    )
+    nbuild_args = nbuild_parser.parse_args()
+
+
+def get_args():
+    """
+    Returns an object holding the value of each command line arguments.
+    It is the return value of `argprase.ArgumentParser.parse_args()`, so refer to `argparse`'s documentation
```
Comment (doom) [inline core/args.py]: Typo: argp**ar**se.ArgumentParser

File: core/args.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        action='count',
+        default=0,
+        help="Make the operation more talkative. Append it multiple times to make it even more talkative."
+    )
+    nbuild_args = nbuild_parser.parse_args()
+
+
+def get_args():
+    """
+    Returns an object holding the value of each command line arguments.
```
Comment (doom) [inline core/args.py]: I think "each" requires a singular, unless you use "each of", in which case it requires a plural.
Also, maybe consider using a plural for "value" too

File: core/cache.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python3.6
+# -*- coding: utf-8 -*-
+"""
+Functions to manipulate the different kind of caches.
```
Comment (doom) [inline core/cache.py]: kind**s**

File: core/cache.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        'install',
+        build.manifest.name,
+        build.semver,
+    )
+
+
+def get_download_cache(build: stdlib.Build):
+    """
+    Returns the location of the cache where the files downloaded by the given build should be placed.
+    This cache is kept across build to avoid downloading the same files over and over.
```
Comment (doom) [inline core/cache.py]: build**s**

File: core/log.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+import termcolor
+from contextlib import contextmanager
+
+log_tab_level = 0
+
+
+@contextmanager
+def pushlog():
+    """
+    Increases the log indentation level by one, making all new lines indented by one more tabulation.
```
Comment (doom) [inline core/log.py]: "making every new line indented by one extra tabulation" ?

File: core/log.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    """
+    global log_tab_level
+
+    indent = '\t' * log_tab_level
+    print(f"{termcolor.colored('[*]', 'blue', attrs=['bold'])} {indent}", *args)
+
+
+def slog(*args, indent=True):
+    """
+    Prints an informative log, prefixed by a green `[+]`.
```
Comment (doom) [inline core/log.py]: "informative" -> "success"

File: nbuild.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        exit(0)
+
+    if core.args.get_args().manifest is None:
+        flog("No path to a build manifest given.")
+        exit(1)
+
+    # Clear environment, inflate a default one
+    os.environ.clear()
+
+    # Target & host architecture
```
Comment (doom) [inline nbuild.py]: What about writing "and" ? :p

**Review by doom — CHANGES_REQUESTED** [review]

**Pull request merged**